cpp/src/fil/common.cuh

```diff
@@ -42,31 +43,33 @@ __host__ __device__ __forceinline__ int forest_num_nodes(int num_trees,
 const int FIL_TPB = 256;
 
 /** base_node contains common implementation details for dense and sparse nodes */
-struct base_node {
+struct base_node : dense_node_t {
   static const int FID_MASK = (1 << 30) - 1;
   static const int DEF_LEFT_MASK = 1 << 30;
   static const int IS_LEAF_MASK = 1 << 31;
-  float val;
-  int bits;
-  __host__ __device__ float output() const { return val; }
-  __host__ __device__ float thresh() const { return val; }
+  __host__ __device__ val_t output() const { return val; }
```

Comment:
What about having a version of `output()` templatized by `leaf_payload_type`? The return type can be templatized as well.